include/proc.h

```diff
@@ -10,37 +10,55 @@
 
 typedef struct thread thread_t;
 typedef struct proc proc_t;
+typedef struct pgrp pgrp_t;
 typedef struct fdtab fdtab_t;
 typedef TAILQ_HEAD(, proc) proc_list_t;
+typedef LIST_HEAD(, pgrp) pgrp_list_t;
+
+/*! \brief Structure allocated per process group.
+ *
+ * Field markings and the corresponding locks:
+ *  (a) all_proc_mtx
+ *  (@) pgrp::pg_lock
+ *  (!) read-only access, do not modify!
+ */
+typedef struct pgrp {
+  mtx_t pg_lock;                /* Process group mutex */
+  LIST_ENTRY(pgrp) pg_link;     /* (a) link on chain of process groups */
+  LIST_HEAD(, proc) pg_members; /* (@) members of process group */
+  pgid_t pg_id;                 /* (!) process group id */
+} pgrp_t;
 
 typedef enum { PS_NORMAL, PS_DYING, PS_ZOMBIE } proc_state_t;
 
 /*! \brief Process structure
  *
  * Field markings and the corresponding locks:
- *  - a: all_proc_mtx
- *  - p: proc_t::p_lock
- *  - @: read-only access
- *  - ~: always safe to access
+ *  (a) all_proc_mtx
+ *  (@) proc_t::p_lock
+ *  (!) read-only access, do not modify!
+ *  (~) always safe to access
  */
 struct proc {
   mtx_t p_lock;               /* Process lock */
   TAILQ_ENTRY(proc) p_all;    /* (a) link on all processes list */
   TAILQ_ENTRY(proc) p_zombie; /* (a) link on zombie process list */
   TAILQ_ENTRY(proc) p_child;  /* (a) link on parent's children list */
-  thread_t *p_thread;         /* (p) the only thread running in this process */
-  pid_t p_pid;                /* (@) Process ID */
-  volatile proc_state_t p_state;  /* (p) process state */
+  thread_t *p_thread;         /* (@) the only thread running in this process */
+  pid_t p_pid;                /* (!) Process ID */
+  LIST_ENTRY(proc) p_pglist;  /* (?) link on p_pgrp->pg_members list */
```

Comment:
`  LIST_ENTRY(proc) p_pglist;  /* (?) link on p_pgrp->pg_members list */
`
Shouldn't it be "(a)"? 

Also why LIST_ENTRY instead of TAILQ_ENTRY as everywhere else in this file?

Reply:
ok, you are right, it should be pgrp_t::pg_lock, but I would note it here

Reply:
done